02-intermediate-ruby/lazy-with-let.md

```diff
@@ -0,0 +1,180 @@
+- DRY up your code with let & before.
+- Explain how `let` is "lazy" in Minitest.
+
+In your Minitest code you will often find yourself creating instances over and over again in each test case like this:
+
+```ruby
+it "must have a name" do
+  player = Scrabble::Player.new("Ada Lovelace")
+  
+  # expectations go here
+end
+```
+
+In most test cases we're creating a new identical instance of `Player` to work with.  This doesn't seem DRY.  
+
+## Solution 1:  Before Blocks
+
+One solution is to use Minitest's `before` block. 
+
+```ruby
+describe Player do
+  before do
+    @player = Scrabble::Player.new("Ada Lovelace")
+  end
+
+  it "must respond to name" do
+    @dplauer.must_respond_to :name
+  end
+  ...
+end
+```
+
+So in the above example the `before` block runs before each test and provides an instance variable `@player` which we can use in our tests.  The problem with before blocks is that they **always** run before each test case.  Even if we don't want to use `@player` in a few test cases, like for instance if we want to try creating a Player with no arguments the before block will still run.  
+
+## Let
+
+Let is another way to DRY up your code.  With Let we create a block which can be referenced in our test cases.
+
+
+```ruby
+# player_spec.rb from Scrabble Project
+
+describe Player do
+  let (:player)  { 						 # <-- The name used
+  	Scrabble::Player.new("Ada Lovelace") # <-- The value returned by 'player'
+  }
+
+
+	it "has a name" do
+		player.name.must_equal "Watson"
+    end
+end
+```
+
+The `let` statement creates a new method with the given name (in this case `player`) that executes the code in the curly braces and caches the result for later use.  That's why it's often called "lazy" the code inside the block doesn't execute until the let is first used, in this case when we do `player.name.must_equal "Watson"`.  So the let runs for each test case that uses it and doesn't run when it isn't needed.  Neat!
+
+### Proof that let is lazy
+
+In this example using `before` the statement "Creating a player named Ada Lovelace" prints twice times showing that before always runs.  
+
+```ruby
+# sample_spec.rb
+
+describe "Player" do
+  before do
+  	puts "Creating a player named Ada Lovelace"
+    @player = Scrabble::Player.new "Ada Lovelace"
+  end
+
+    describe "initialize method" do
+        it "New Players initialize with a name" do
+            @player.name.must_equal "Ada Lovelace"
+        end
+
+        it "Throws an ArgumentError if created without a name" do
+          proc {
+            Scrabble::Player.new
+          }.must_raise ArgumentError
+        end
+    end
+end
+```
+Output:
+
+```bash
+$  ruby specs/sample_spec.rb
+Run options: --seed 62527
+
+# Running:
+	
+Creating a player named Ada Lovelace
+.Creating a player named Ada Lovelace
+.
+
+Finished in 0.011018s, 272.2817 runs/s, 272.2817 assertions/s.
+
+2 runs, 2 assertions, 0 failures, 0 errors, 0 skips
+```
+
+But if we use let instead:
```

Comment:
I think I'd be good to add headings above the `before` example and the `let` example so it's a bit easier to follow along with these sections